### ISSUE TITLE
pytest.warns() as a context manager is difficult to check multiple warnings

### ISSUE THREAD

**Issue by s-t-e-v-e-n-k** [pytest-dev/pytest#12456]
#### What's the problem this feature will solve?

Using `pytest.warns()` as a context manager will with pytest 8.x, now emit all warnings, which is excellent, since now you can check all warnings that code produces, rather than matching on only one warning and it swallowing the rest. Nesting `pytest.warns()` will help with multiple warnings, but this quickly becomes very unwieldy. 

#### Describe the solution you'd like

Consider the following test snippet:

```
with pytest.warns(FeatureDeprecated) as all_warnings:
    deprecated_feature()
```

If this raises multiple warnings, `all_warnings` now gives all of them, but then no real way to assert all of them are the expected warnings, requiring users to iterate over them and write a bunch of code to check them. This is already supported with only one warning, using the `match` parameter, but it doesn't seem to help here.

I'm envisioning something like:

```
expected_warnings = [FeatureDeprecated("Foo"), FeatureDeprecated("Bar")]
all_warnings.match_all(expected_warnings)
```

Which would iterate over all warnings passed in, and would `assert` if there are extra warnings or some did not match, and also allow users to construct warnings with regular expressions to have them matched against, like `match` does now.

#### Alternative Solutions

I have written code like the below, which just feels too wordy.

```
        with pytest.warns(PydanticDeprecatedSince20) as all_warnings:

            class Config(BaseConfig):
                pass

        assert len(all_warnings) == 2
        expected_warnings = [
            'BaseConfig is deprecated. Use the `pydantic.ConfigDict` instead',
            'Support for class-based `config` is deprecated, use ConfigDict instead',
        ]
        assert [w.message.message for w in all_warnings] == expected_warnings
```

#### Additional context

I have yet to push up my pydantic PR, but wanted to get some feedback if this is a terrible or a good idea.

**Comment by mivanit** [comment]
If I'm not mistaken this feature is already present: https://docs.pytest.org/en/8.2.x/how-to/capture-warnings.html#recwarn

(found myself here because I was having the same problem)

**Comment by Zac-HD** [comment]
> Nesting `pytest.warns()` will help with multiple warnings, but this quickly becomes very unwieldy.

From Python 3.10, [you can have as many context managers as you want with only a single level of indentation](https://docs.python.org/3/whatsnew/3.10.html#parenthesized-context-managers)!
